### ISSUE TITLE
Adding route constraints to versioning doesn't pick them back up on VersionedApiExplorer 

### ISSUE THREAD

**Issue by cberthold** [Microsoft/aspnet-api-versioning#275]
I have a custom route constraint that I've added that subdivides our application by picking up a custom attribute on the controller and matching on incoming requests by overriding `DefaultDirectRouteProvider` and adding the constraint to each controller based on each `HttpControllerDescriptor` coming in.  I've got that working and performs wonderfully with api versioning for incoming requests.  Where I'm struggling is subdividing Swagger based on that same premise.  I've added the ConstraintMap to the `DefaultInlineConstraintResolver` like so 
`var constraintResolver = new DefaultInlineConstraintResolver()
            {
                ConstraintMap =
                {
                    ["apiVersion"] = typeof(ApiVersionRouteConstraint),
                    [AgentConstants.RouteConstraintKey] = typeof(AgentRouteConstraint),
                }
            };
            var agentRouteProvider = new AgentDirectRouteProvider();
            config.MapHttpAttributeRoutes(constraintResolver, agentRouteProvider);
`

It seems like I could override GetGroupName of VersionedApiExplorer and return the string and possibly break it back out in the groups on the swagger side and maybe use an IOperationFilter but I'd like to achieve somewhat of the same goal to have a route constraint applied to the list of applicable routes.  Ideally I can use that same constraint to only show the values in Swagger UI.  Some guidance would be appreciated.

**Comment by commonsensesoftware** [comment]
Sorry, but I'm not sure follow. What would the resultant URLs look like? I'm not sure what you mean by _subdivide_. If you're able to outline a basic illustration or description of what you expect success to look like, it would be helpful.

**Comment by cberthold** [comment]
Sure.  So I'm actually creating duplicate controllers in a different namespace so that I have Agent "A" in namespace "A" and Agent "B" in namespace "B".  I'm using a route constraint that looks up the expected agent using a header "api-agent" to signify which agent namespace to apply to.  The attribute is being applied to a base ApiController which is inconsequential to the implementation.  Then the [ApiVersion("1.0")] Tags are being applied.  It is all done with Route/Prefix attributes.  No mixing.  My expected result is that I can use the same route constraint against the swagger api to only give the versions available to that "api-agent" tag.  In my route constraint, I have made it so that can work via header or URI.  My hope was that when VersionedApiExplorer gets called it would being calling the route constraints in the same way to match the controllers.  Right now it seems like it picks the first matching route in each case.  I think I could subdivide by overriding GetGroupName in VersionedApiExplorer and then parsing it back out in the api descriptions but that feels ucky and was hoping for the cleaner solution. My ideal is that the same route constraint just evaluates the request coming into swagger and calls that same route constraint.  Resultant URLs would look like localhost:8000/v1/swagger?api-agent=A or B but what that returns would be dependent on A or B.

**Comment by commonsensesoftware** [comment]
I see. This seems like it's more on the Swagger side. In Swashbuckle, I'm not sure how you get in front of the request. I don't think even the group name helps because you have to wait for an incoming request to know what the client is requesting. If you can somehow get to the incoming request, it shouldn't be a big deal change things. The API Explorer works completely off of known data. There's nothing used from an incoming request.

It _sounds_ like what you need is a custom Swagger generator. You can probably extend a built-in one. You can probably do it with a custom **IDocumentFilter** or **IOperationFilter**. Otherwise, it seems you cannot use `configuration.EnableSwagger` and you'll have to write you own handler that does the filtering the way that [SwaggerDocsHandler.cs](https://github.com/domaindrivendev/Swashbuckle/blob/master/Swashbuckle.Core/Application/SwaggerDocsHandler.cs) works. That seems to be the only way to get to the corresponding request.

**Comment by cberthold** [comment]
That's kind of what I was figuring.  I went down that path and GetRootUrl was not available so it started to snowball.  With GetGroupName I can make it sort of work because instead of returning V1.0 I can return AgentA-V1.0 and that would give the extra subdivision and I can break it back out in MultipleVersions.  The reason I thought it might be possible with **VersionedApiExplorer** is because when I looked at the way the **ApiVersionRouteConstraint** worked in the match I saw:

`
if ( routeDirection == UrlGeneration )
{
    return !IsNullOrEmpty( value );
}
`

That made me believe there must be a way that URL's where being generated and somehow I could back feed the appropriate Agents in a similar way.  I'm so close I can taste it.  Enabling this scenario allows for a very clean separation for various agents but the "copy-paste" scenario without great effort.  Really appreciate the hard work on api-versioning.  It has enabled a really clean way to handle versions and nearly everything I could've wanted has been implemented already.

**Comment by cberthold** [comment]
So I've made a little progress on the **VersionedApiExplorer** and figured out that I can override **ExploreDirectRouteControllers** and add metadata to the returned collection of **VersionedApiDescription** in the Properties of that for each mapped route.  I was then going to  extend **InitializeApiDescriptions** to return an ApiDescriptionGroupCollection representing the multiple versions v1 for Agent A and v1 for Agent B.  I then noticed that the Flattening of the versions is also tossing the duplicates as a result.  I'm trying to avoid copying and making my own complete version of VersionApiExplorer.  By doing that I could then create my own version of UseSwagger which adds in the route constraints and ties that using the **IDocumentFilter* like you suggested.  Right now it seems like if I replace the **ApiDescriptionComparer** with one that knows about the agent routes so it stops flattening out all of "duplicates" and then copying the ApiVersion to an overridden version that will also contain the agent information and break it back out in **InitializeApiDescriptions**  that will get me over to swagger and now its completely my problem.  Am I missing anything in there?

**Comment by commonsensesoftware** [comment]
Sounds like you're on the right track. Overriding **InitializeApiDescriptions** would probably be the simplest. Call the base implementation and then apply the filtering at the end. The default behavior bucketizes by API version so you don't want any duplicates there.

In Web API, the **IApiExplorer** doesn't support the concept of _grouping_. I had to _beat_ it into submission. Other frameworks like Swashbuckle and NSwag depend on the **IApiExplorer** contract. The API versioning flavor flattens the grouped list to adhere to the contract. At least in Swashbuckle, a filtering callback is defined to determine whether a descriptor matches. I provided an extension method to get the group name so it's easy to integrate with.

Another approach to subclassing would be to use the _Decorator_ pattern to apply the filter to the results. Creating the instance would look something like:

```c#
var apiExplorer = new TenantSpecificApiExplorer(
    "tenant",
     new VersionedApiExplorer( config, options ) );
```

You implementation just needs to call the inner **IApiExplorer** and then filter the results. I'll leave it up to you as to whether that's better than subclassing. It _should_ be easier to test. ;)

**Comment by cberthold** [comment]
So I started to think of a suggestion (adding a version factory) and then it hit me.  As I went to describe the feature and discover for it I found the part I was missing. It was that the **ApiVersionAttribute** was in fact the provider of **ApiVersion**.  I missed it in the discovery phase that the **ApiVersionModel** was using the **ApiVersionAttribute** as the factory or provider for **ApiVersion**.  If I extend both the attribute(s) and the ApiVersion that will get me down through the **IApiExplorer** without having to redo all of the comparison.  Instead I just need to override the comparison for ApiVersion and bingo bango I have a different version automatically segmented.  I still have to figure out Swashbuckle. The groupname is still the correct concept.  I want V1, V2, V3 but when I pass in Agent=A I want to narrow the results to that segment.  I really appreciate the help, sometimes it helps to talk through stuff.  It was like pair programming :) Also, great work on this feature.  Literally everything to extend it is there once you understand how the model was put together.  I really learned a lot.

**Comment by commonsensesoftware** [comment]
Remember that _API version-neutral_ does not mean _no API version_; it's exactly the opposite. It means _all, any, or no API version_. If a controller is **IApiVersionNeutral**, then no API versions will be provided. The API Explorer adds API version-neutral actions to every discovered API version. Also keep in mind that it's a mistake to have a duplicate, API version-neutral route across namespaces. If a controller is API version-neutral, _"...there can be only one!"_ ;)  There's no way to make a selection otherwise because the choices are ambiguous.

**PopulateApiVersionParameters** is just adding the API version parameter, wherever that may be. The API Explorer doesn't know how or where the API version parameter is. While it may seem like it _should_ know that, it would limit it's design and usefulness to only API versioning methods it knows how to deal with. Instead, it relies on the **IApiVersionParameterSource** which knows exactly how to describe the parameter to be added. Today, this is still limited to query, header, path, or media type, but that will probably cover everything for all time. The **IApiVersionReader** is, therefore, also a **IApiVersionParameterSource**. This could get a lot more complicated for the API Explorer because an **IApiVersionReader** can use the _Decorator_ pattern to compose multiple styles. Each style can add a parameter. For example, an API might allow for a parameter in the query string and header; thereby resulting in two parameters. As long as each composed **IApiVersionReader** and, hence, **IApiVersionParameterSource** can say what the name of the parameter is and where it's location is, the API Explorer knows how to build an appropriate **ApiParameterDescription**. This is all necessary because the API version is not formally defined in your API model anywhere. Ultimately, **PopulateApiVersionParameters** does not provide an API version for **IApiVersionNeutral** actions because which API version would you provide? Furthermore, providing no API version is still valid. I hope that clarifies how it works.

**Comment by cberthold** [comment]
It does clarify it a ton.  As with any "HACK" it turns out that worked out really well.  So I use **PopulateApiVersionParameters** which gets to see even the Neutral ones.  Leaving my extra attribute on my base controllers for the segmentation was definitely the way to go for that exact reason you mention.  So given that, when the **ApiDescription** comes across I merely need to convert it back to **VersionedApiDescription** and then convert and replace the **ApiVersion** with my extended **AgentApiVersion** with my additional metadata and overrides for equality and hashing.  The **ApiDescriptionComparer** gets replaced for extended implementation doing all of those same things using the AgentAtttribute provider I created that stores them on the Properties.  That prevented the complete flattening of all the V1-A and V1-B urls that look exactly the same. To go from V1, V2 to V1-A, V1-B, V2-A,V2-B, I did override **InitializeApiDescriptions** and looped the groups/descriptions and just started a new GroupCollection and GetOrAdd with the ApiVersion which was now the **AgentApiVersion** so it split it automatically through the hash/equality overrides. Like you mentioned GroupName was the secret to bridging the version over to Swashbuckle.  I only needed it to be unique in order to get over to Swashbuckle.  Once over there I redescribed the routes and replaced information back the way I wanted it.  The trick was to add Swagger/UI routes for each agent passing that into the versionSupportResolver portion of  swagger.MultipleApiVersions to narrow to that particular agent.  I hit a couple bugs adding the swagger routes multiple times, mostly had to remove the route it creates and add it back with a unique name.  Again I really appreciate all the help.  I'm going to see if I can put some PR's in for the stuff I hit.  I think what is really missing on the Swashbuckle side is a request metadata provider.  I think since I'm using autofac I could've resolved the HttpRequestMessage off of their CustomProvider but that would've been too ucky.  Having metadata about the request I could've done it all with document/operation filters and had a much cleaner and lightweight solution.  Thanks again for your time.  Hope this conversation helps someone out at some point.  Have a great weekend.
